File: tests/pytest/query/query.py
###################################################################
#           Copyright (c) 2016 by TAOS Technologies, Inc.
#                     All rights reserved.
#
#  This file is proprietary and confidential to TAOS Technologies.
#  No part of this file may be reproduced, stored, transmitted,
#  disclosed or used in any form or by any means other than as
#  expressly provided by the written permission from Jianhui Tao
#
###################################################################

# -*- coding: utf-8 -*-

import sys
import taos
from util.log import tdLog
from util.cases import tdCases
from util.sql import tdSql
from util.dnodes import tdDnodes

class TDTestCase:
    def init(self, conn, logSql, replicaVar = 1):
        tdLog.debug("start to execute %s" % __file__)
        tdSql.init(conn.cursor(), logSql)

        self.ts = 1538548685000

    def bug_6387(self):
        tdSql.execute("drop database if exists db")
        tdSql.execute("create database if not exists db")
        tdSql.execute("use db")
        tdSql.execute("create table test(ts timestamp, c1 int) tags(t1 int)")
        for i in range(5000):
            sql = "insert into t%d using test tags(1) values " % i
            for j in range(21):
                sql = sql + "(now+%ds,%d)" % (j ,j )
            tdSql.execute(sql)
#        tdSql.query("select count(*) from test interval(1s) group by tbname")
#        tdSql.checkData(0,1,1)

    def run(self):
        tdSql.prepare()

        print("==============step1")
        tdSql.execute(
            "create table if not exists st (ts timestamp, tagtype int) tags(dev nchar(50))")
        tdSql.execute(
            'CREATE TABLE if not exists dev_001 using st tags("dev_01")')
        tdSql.execute(
            'CREATE TABLE if not exists dev_002 using st tags("dev_02")')

        print("==============step2")

        tdSql.execute(
            """INSERT INTO dev_001(ts, tagtype) VALUES('2020-05-13 10:00:00.000', 1),
            ('2020-05-13 10:00:00.001', 1)
             dev_002 VALUES('2020-05-13 10:00:00.001', 1)""")

        tdSql.query("select * from db.st where ts='2020-05-13 10:00:00.000'")
        tdSql.checkRows(1)

#        tdSql.query("select tbname, dev from dev_001") 
#        tdSql.checkRows(1)
#        tdSql.checkData(0, 0, 'dev_001')
#        tdSql.checkData(0, 1, 'dev_01')

        tdSql.query("select tbname, dev, tagtype from dev_001") 
        tdSql.checkRows(2)
        tdSql.checkData(0, 0, 'dev_001')
        tdSql.checkData(0, 1, 'dev_01')
        tdSql.checkData(0, 2, 1)
        tdSql.checkData(1, 0, 'dev_001')
        tdSql.checkData(1, 1, 'dev_01')
        tdSql.checkData(1, 2, 1)

        ## test case for https://jira.taosdata.com:18080/browse/TD-2488
        tdSql.execute("create table m1(ts timestamp, k int) tags(a int)")
        tdSql.execute("create table t1 using m1 tags(1)")
        tdSql.execute("create table t2 using m1 tags(2)")
        tdSql.execute("insert into t1 values('2020-1-1 1:1:1', 1)")
        tdSql.execute("insert into t1 values('2020-1-1 1:10:1', 2)")
        tdSql.execute("insert into t2 values('2020-1-1 1:5:1', 99)")
        
        tdSql.query("select count(*) from m1 where ts = '2020-1-1 1:5:1' ")
        tdSql.checkRows(1)
        tdSql.checkData(0, 0, 1)

        tdDnodes.stop(1)
        tdDnodes.start(1)

        tdSql.query("select count(*) from m1 where ts = '2020-1-1 1:5:1' ")
        tdSql.checkRows(1)
        tdSql.checkData(0, 0, 1)

        ## test case for https://jira.taosdata.com:18080/browse/TD-1930
        tdSql.execute("create table tb(ts timestamp, c1 int, c2 binary(10), c3 nchar(10), c4 float, c5 bool)")
        for i in range(10):
            tdSql.execute("insert into tb values(%d, %d, 'binary%d', 'nchar%d', %f, %d)" % (self.ts + i, i, i, i, i + 0.1, i % 2))
        
        tdSql.error("select * from tb where c2 = binary2")
        tdSql.error("select * from tb where c3 = nchar2")

        tdSql.query("select * from tb where c2 = 'binary2' ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c3 = 'nchar2' ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c1 = '2' ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c1 = 2 ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c4 = '0.1' ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c4 = 0.1 ")
        tdSql.checkRows(1)

        tdSql.query("select * from tb where c5 = true ")
        tdSql.checkRows(5)

        tdSql.query("select * from tb where c5 = 'true' ")
        tdSql.checkRows(5)

        # For jira: https://jira.taosdata.com:18080/browse/TD-2850
        tdSql.execute("create database `Test` ")
        tdSql.execute("use `Test` ")
        tdSql.execute("create table TB(ts timestamp, `Col1` int) tags(`Tag1` int)")
        tdSql.execute("insert into Tb0 using tb tags(1) values(now, 1)")
        tdSql.query("select * from tb")
        tdSql.checkRows(1)

        tdSql.query("select * from tb0")
        tdSql.checkRows(1)

        #For jira: https://jira.taosdata.com:18080/browse/TD-6387
        #self.bug_6387()
        

    def stop(self):
        tdSql.close()
        tdLog.success("%s successfully executed" % __file__)


tdCases.addWindows(__file__, TDTestCase())
tdCases.addLinux(__file__, TDTestCase())

File: source/taos-internal/enterprise/tests/test/cdriver/query.c
// gcc query.c -o ../../../build/bin/query -g -I../../inc -L../../../build/lib -ltsclient -lttaos -ltutil -lpthread 

/*******************************************************************
 *           Copyright (c) 2017 by TAOS Technologies, Inc.
 *                     All rights reserved.
 *
 *  This file is proprietary and confidential to TAOS Technologies. 
 *  No part of this file may be reproduced, stored, transmitted, 
 *  disclosed or used in any form or by any means other than as 
 *  expressly provided by the written permission from Jianhui Tao
 *
 * ****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/time.h>
#include <string.h>
#include <assert.h>

#include "taos.h"
#include "tsclient.h"
#include "tutil.h"

void taos_error(TAOS *con)
{
  fprintf(stderr, "TSDB error: %s\n", taos_errstr(con));
}

void query(TAOS* con, char* q) {
  struct timeval systemTime;
  gettimeofday(&systemTime, NULL);
  int64_t st = systemTime.tv_sec*1000000 + systemTime.tv_usec;

  if (taos_query(con, q) != 0 ) {
    taos_error(con);
    exit(1);
  }

  TAOS_RES* result = taos_use_result(con);

  if (result == NULL)
    taos_error(con);

  TAOS_ROW row;

  int32_t numOfRows = 0;
  int num_fields = taos_field_count(con);
  TAOS_FIELD *fields = taos_fetch_fields(result);
  char   temp[256] = {0};

  while ((row = taos_fetch_row(result)))
  {
      for(int i = 0; i < num_fields; i++)
      {
        switch( fields[i].type ) {
          case TSDB_DATA_TYPE_TINYINT:
            sprintf(temp + strlen(temp), "%d ", *((char *)row[i]));
                break;
          case TSDB_DATA_TYPE_SMALLINT:
            sprintf(temp + strlen(temp), "%d ", *((short *)row[i]));
                break;
          case TSDB_DATA_TYPE_INT:
            sprintf(temp + strlen(temp), "%d ", *((int *)row[i]));
                break;
          case TSDB_DATA_TYPE_BIGINT:
            sprintf(temp + strlen(temp), "%lld ", *((int64_t *)row[i]));
                break;
          case TSDB_DATA_TYPE_FLOAT:
            sprintf(temp + strlen(temp), "%f ", *((float *)row[i]));
                break;
          case TSDB_DATA_TYPE_DOUBLE:
            sprintf(temp + strlen(temp), "%lf ", *((double *)row[i]));
                break;
          case TSDB_DATA_TYPE_BINARY:
            sprintf(temp + strlen(temp), "%s ", (char *) row[i]);
                break;
          case TSDB_DATA_TYPE_TIMESTAMP:
            sprintf(temp + strlen(temp), "%lld ", *((time_t *)row[i]));
                break;
          default:
            break;
        }
    }
    fprintf(stdout, "%s\n", temp);
    memset(temp, 0, sizeof(temp)/sizeof(temp[0]));
    numOfRows++;
  }


  gettimeofday(&systemTime, NULL);
  int64_t et = systemTime.tv_sec*1000000 + systemTime.tv_usec;
  printf("%.3f mseconds to retrieve %d data points\n", (et-st)/1000.0, numOfRows);
//  assert(numOfRows == 200);
  taos_free_result(result);
}

int main(int argc, char *argv[]) {
  TAOS *con;
  struct timeval systemTime;
  int64_t st, et;
  char qstr[128], db[128] = "test";
  char table[20] = "tm0";
  int order = 0;
  int save = 1;
  int numOfRows;
  TAOS_RES *result;
  FILE *fp;

  if (argc == 1) {
    printf("usage: %s db cfg\n", argv[0]);
    exit(0);
  }

  //t1 db1 0 1 . / cfg
  if (argc >= 2) strcpy(db, argv[1]);
//  if (argc >= 3) strcpy(db, argv[2]);
//  if (argc >= 4) order = atoi(argv[3]);
//  if (argc >= 5) save = atoi(argv[4]);
  if (argc >= 3) strcpy(configDir, argv[2]);

  con = taos_connect(tsMasterIp, tsDefaultUser, "1", NULL, 0);
  if (con == NULL) {
    taos_error(con);
    exit(1);
  }

  sprintf(qstr, "create database %s", db);
  taos_query(con, qstr);

  sprintf(qstr, "use %s", db);
  taos_query(con, qstr);

//  taos_query(con, "drop table tt");

//  taos_query(con, "drop table testmetric");
//  taos_query(con, "create table testmetric(ts timestamp, k int) tags(a int)");
//  taos_query(con, "create table tt using testmetric tags(1)");
//  taos_query(con, "insert into tt values(now, 1)");

  for(int32_t i =0; i<100; ++i) {
    printf("%d-----------------------------------------------------\n\n\n", i);

    //error test case: @@@@@@!!!!
    query(con, "select count(*), avg(k), sum(y), count(y), avg(y) from m1 group by a");
    query(con, "select count(*), first(k), last(k), min(f),max(f),sum(y),avg(y) from m1 group by c");
  }
  taos_close(con);
}


